### ISSUE TITLE
Add MAD (Median Absolute Deviation) to base functions.

### ISSUE THREAD

**Issue by DylanLukes** [pola-rs/polars#10578]
### Problem description

Currently there's a bit of an asymmetry in the provided statistical functions:

|  | Non-Robust | Robust |
| --- | --- | --- |
| **Central Tendency** | ✅ `mean` |  ✅ `median` |
| **Dispersion** |  ✅ `std` | ❌ `mad` |

For parity, it would be nice to include the robust estimator of dispersion. It is defined as the median of the absolute differences between each point and the median. 

**N.B:**
The value of the MAD itself is not an estimator of the standard deviation on its own, but if can be made into one by scaling. Generally speaking, the choice of `scale` depends on the distribution. For normally distributed data, dividing by the inverse of the standard normal quantile function at 0.75 (~0.67449) or equivalently multiplying by 1.4826 approximates the standard deviation. [Scipy affords this functionality with the special value `"normal"` for the `scale` parameter. It also allows for "t".](https://github.com/scipy/scipy/blob/44e31b1eb202ef91caad059952ba268d2610a60b/scipy/stats/stats.py#L8099)

**Other libraries:**
- Numpy does not provide an equivalent function
- Scipy provides `scipy.stats.median_abs_deviation`.
- Pandas used to (`df.mad) but it was incorrect (returned the _mean_ absolute deviation) and has since been deprecated.

**Example minimal implementation:**

```
@pl.api.register_expr_namespace("example")
class Example:
    def mad(self, scale: float | Literal["normal"] = 1.0) -> pl.Expr:
        if scale is "normal":
            scale = ... # ❗️
        return scale * (self.expr - self.expr.median()).abs().median()
```

The main counterargument I can see is that calculating the scale parameter (❗️) involves evaluating either a magic constant or evaluating the CDF of a distribution at runtime (like SciPy), which seems out of scope for Polars. Scipy's implementation delegates to [`ndtr`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.special.ndtr.html) which is [provided by the Cephes special function C library](https://github.com/scipy/scipy/blob/7315b9fc434e5a8a188ab107cf949c12e1465276/scipy/special/cephes/ndtr.c#L201). Luckily, the special functions involved are limited (error function, it's complement, and polynomial evaluation), and simple to translate into Rust, so this should not require any dependency. 

If this is something that could be considered for inclusion (including Rust translations for `ndtr`, `erf`, `erfc`, `polevl`/`p1evl`), I'd be happy to prepare a pull request with an implementation.

**Comment by zundertj** [comment]
Three thoughts I have:

1. It is, per your example, easy enough to code using the existing expression API if you use a simple constant, so the argument for adding to Polars would be purely about user convenience, not performance, right? So the debate would be about how niche it is for it to (not) belong in Polars. I sometimes think this sort of thing would be better of in a separate package that implements a large number of these statistics; would be interested to learn if there are any such packages already.
2.We could start with using the constant first? Are there situations where the difference is large versus using the actual CDF to practically matter, given the primary reason for using mad over std to measure std is robustness to outliers.
3. Using `mad` for both Mean and Median Absolute Deviation always confuses me, especially as different librariers use `mad` for either as you point out. Hence I would strongly prefer `median_abs_dev` (or some other longer version) over `mad` as the name. Not as nice and short as `std`, but I prefer to be clear here.

**Comment by DylanLukes** [comment]
> The argument for adding to Polars would be purely about user convenience, not performance, right?

Yes, the main argument is convenience (with a dash of appeal to "completeness" as noted in the table). In my own code, I've used an `Expr` namespace extension. (Sadly, this doesn't play very well with in a strictly typed context...)

> I sometimes think this sort of thing would be better off in a separate package that implements a large number of more niche statistics/transformations (am thinking z-scores, percent rankings, mean absolute deviation, OLS slope, etc); would be interested to learn if there are any such packages already.

I haven't found one, but I was considering implementing a simple one for my own use on top of SciPy. Of course, that means round-tripping through `numpy`, which presents some issues (e.g. `to_numpy` is only guaranteed zero-copy in the absence of nulls, black-box to optimization akin to `apply`). But, at least as a starting point, it would be nice and easy likely enough. (Aside: [1])

> Using mad for both Mean and Median Absolute Deviation always confuses me.

It's really just Pandas that dropped the ball here 😅. I haven't seen `mad` used for _mean_ absolute deviation anywhere else. To add another data point, [the standard name in R is `mad` as well](https://www.rdocumentation.org/packages/stats/versions/3.6.2/topics/mad). 

> Not as nice and short as std, but I prefer to be clear here over saving a couple of characters.

There is a more general term, [average absolute deviation (AAD)](https://en.wikipedia.org/wiki/Average_absolute_deviation), which encompasses the same form for any measure of central tendency (mean, arithmetic mean, median, mode, ...). Could just have an `aad` that is parametrized by a `Callable` or `Expr`. Then `mad` is just a special case for `measure="median"`.

---

[1] This is a bit of an aside, but one roadblock I haven't figured out how to overcome is that the use of the `pl.api.register_*` family is rather incompatible with strict typing using MyPy. For example, you can't tell the type-checker that `pl.Expr.foo` is a `NameSpace`, not to mention that it's a `NameSpace` that corresponds to a certain extension class with particular methods. I think the only realistic way forward here is a MyPy plugin, sadly.  

This presents an issue for my own little local assortment of statistics, but also for implementing a proper statistics namespace in a separate package as previously mentioned.

**Comment by yanxiu0614** [comment]
```
def clip(series):
    median = series.quantile(0.5)
    new_median = ((series - median).abs()).quantile(0.50)
    max_range = median + 3*new_median
    min_range = median - 3*new_median
    return series.clip(min_range,max_range)
 _pdf=_pdf.with_columns(
                    pl.col("xx").apply(clip).over("yy").alias("xx"))
```

In fact, this piece of code is very slow, and it might also be due to my code being incorrect. If this new function can speed things up, it's very much anticipated.

**Comment by zundertj** [comment]
@yanxiu0614 : if you want speed, `apply` is not the way to go. See also the warning in the [docs](https://pola-rs.github.io/polars/py-polars/html/reference/expressions/api/polars.Expr.apply.html#polars.Expr.apply).

I would try something alone these lines (NOTE: havent tested, do your own testing!)
```python
# Expr.clip does not support expressions, make our own
def clip_expr(x: pl.Expr, lower: pl.Expr, upper: pl.Expr):
     return pl.when(x<lower).then(lower).when(x>upper).then(upper).otherwise(x)

def clip_mad(x: pl.Expr):
     xm = x - x.median()
     mad = xm.abs().median()
     lower = x.median() - 3*mad
     upper = x.median() + 3*mad
     return clip_expr(x, lower, upper)

_pdf = _pdf.with_columns(clip_mad(pl.col("xx")).over("yy").alias("xx"))

```

**Comment by yanxiu0614** [comment]
@zundertj Thank you very much for your correction. Not using the apply method has reduced the time I spent by half, but in reality, it's still slow. I use Polars for performing a large number of complex computations. Without using 'mad', each iteration takes around 0.03 seconds. However, with the 'mad' step now included, it takes about 0.1 seconds. So, I still hope that if this function can be implemented, it will have a significant acceleration effect.

**Comment by Wainberg** [comment]
I also think median absolute deviation is sufficiently common to be worth including (and is more useful than mean absolute deviation).

polars already has a function for [exponentially-weighted moving average](https://docs.pola.rs/py-polars/html/reference/series/api/polars.Series.ewm_mean.html), which is much more obscure!

**Comment by alexander-beedie** [comment]
> polars already has a function for [exponentially-weighted moving average](https://docs.pola.rs/py-polars/html/reference/series/api/polars.Series.ewm_mean.html), which is much more obscure!

Not in finance ;)

**Comment by Wainberg** [comment]
Ok touché - "exponentially weighted moving average" has 330k Google results and "median absolute deviation" has 275k.

**Comment by al6x** [comment]
It would be nice to have this function built in, as doing it by hand is not easy (calculating rolling median abs dev with brute force is trivial, but making it fast and efficient is not easy).

As for the naming - yes 'mad' is confusing (is it mean abs deviation or median abs deviation?).
